src/adapter/src/coord/timestamp_selection.rs

```diff
@@ -252,4 +181,90 @@ impl<S: Append + 'static> Coordinator<S> {
             Timestamp::MAX
         }
     }
+
+    fn evaluate_when(
+        &self,
+        mut timestamp: MirScalarExpr,
+        session: &Session,
+    ) -> Result<mz_repr::Timestamp, AdapterError> {
+        let temp_storage = RowArena::new();
+        prep_scalar_expr(self.catalog.state(), &mut timestamp, ExprPrepStyle::AsOf)?;
+        let evaled = timestamp.eval(&[], &temp_storage)?;
+        if evaled.is_null() {
+            coord_bail!("can't use {} as a mztimestamp for AS OF", evaled);
+        }
+        let ty = timestamp.typ(&[]);
+        Ok(match ty.scalar_type {
+            ScalarType::MzTimestamp => evaled.unwrap_mztimestamp(),
+            ScalarType::Numeric { .. } => {
+                let n = evaled.unwrap_numeric().0;
+                n.try_into()?
+            }
+            ScalarType::Int16 => i64::from(evaled.unwrap_int16()).try_into()?,
+            ScalarType::Int32 => i64::from(evaled.unwrap_int32()).try_into()?,
+            ScalarType::Int64 => evaled.unwrap_int64().try_into()?,
+            ScalarType::UInt16 => u64::from(evaled.unwrap_uint16()).into(),
+            ScalarType::UInt32 => u64::from(evaled.unwrap_uint32()).into(),
+            ScalarType::UInt64 => evaled.unwrap_uint64().into(),
+            ScalarType::TimestampTz => evaled.unwrap_timestamptz().timestamp_millis().try_into()?,
+            ScalarType::Timestamp => evaled.unwrap_timestamp().timestamp_millis().try_into()?,
+            _ => coord_bail!(
+                "can't use {} as a mztimestamp for AS OF",
+                self.catalog.for_session(session).humanize_column_type(&ty)
+            ),
+        })
+    }
+
+    fn generate_error_msg(
```

Comment:
Because we're on the coord struct, this fn name should be more descriptive that it's about a timestamp error. Maybe `fn generate_timestamp_not_valid_error_msg`.

Reply:
Just updated